### ISSUE TITLE
matplotlib crashes on OS X when saving to JPEG and then displaying the plot

### ISSUE THREAD

**Issue by netheril96** [matplotlib/matplotlib#4194]
Here is the minimal code to produce the crash on OS X Yosemite, with Python 2.7.9 compiled by homebrew.

```python
# -*- coding: utf-8 -*-
import numpy
import matplotlib.pyplot as plt
plt.plot(numpy.linspace(0, 1, 1000), numpy.random.randn(1000))
plt.savefig('test.jpg')
plt.show()
```

The error message is

```
Assertion failed: (s->stack->next != NULL), function CGGStackRestore, file Context/CGGStack.c, line 77.
```

The crash does not happen if the `savefig` and `show` call are swapped, or when saving to `png`/`pdf`/`svg` file. It only happens when `pyplot` is called in this sequence, and when the format is `jpg`.

**Comment by tacaswell** [comment]
What backend are you using?  I suspect the OSX backend from the error.   Is that the full error message you get?

I can not reproduce this on py3 + linux + qtagg

attn @mdehoon

**Comment by netheril96** [comment]
@tacaswell 

Yes, it is the MacOSX backend, and the error message is all that I get.

If I manually specify the `agg` backend, it works normally as well. So this is a bug specific to OSX backend.

**Comment by jenshnielsen** [comment]
Which Matplotlib version do you use? I could not reproduce this on master (1.5.dev1)

**Comment by netheril96** [comment]
@jenshnielsen 

matplotlib (1.4.3)
Python 2.7.9
Mac OS X backend

**Comment by mdehoon** [comment]
I could not reproduce this with Python 2.7.9 and matplotlib 1.4.3 on Mac OS X 10.9 or 10.10.
Can you try changing backend_macosx.py to return immediately from _print_bitmap ?
So
```python
        def _print_bitmap(self, filename, *args, **kwargs):
            return
```
Then, with some luck, we can find out exactly where the problem is occurring.

**Comment by netheril96** [comment]
@mdehoon 

The error message goes away, and obviously the `jpg` is never created.

I took the liberty of poking around in that function, and the error occurs in the call

`self.write_bitmap(filename, width, height, dpi)`

Even weirder is that running this script in `ipdb` line by line does not produce any error, and does produce a correct `jpeg` file.

**Comment by mdehoon** [comment]
I am guessing that these two lines:
```
    [image release];
    [resizedImage release];
```
in the function FigureCanvas_write_bitmap in _macosx.m are causing the problem. Can you try moving these two lines down to just before
```
    [pool release];
```
, recompile, and see if the problem still occurs?

**Comment by netheril96** [comment]
@mdehoon 

The problem still occurs after the change and recompile/reinstallation.

Deleting that two release calls doesn't help, either.

**Comment by netheril96** [comment]
@mdehoon 

I also tried recompiling python as well as installing matplotlib from this git repository. The assertion failure is still there.

**Comment by mdehoon** [comment]
Sorry, there is not much I can do as I cannot reproduce this error. I would suggest to remove more of the code in FigureCanvas_write_bitmap and see if the problem disappears at some point.

**Comment by mdboom** [comment]
Possibly fixed by #6178, but since we're having trouble reproducing, it's hard to say.
